poolakey/src/main/java/ir/cafebazaar/poolakey/billing/connection/ReceiverBillingConnection.kt

```diff
@@ -0,0 +1,411 @@
+                onQueryPurchaseReceived(extras)
+            }
+        }
+    }
+
+    private fun isPurchaseTypeSupported() {
+        getNewIntentForBroadcast().apply {
+            action = ACTION_BILLING_SUPPORT
+        }.run {
+            sendBroadcast(this)
+        }
+    }
+
+    override fun consume(purchaseToken: String, callback: ConsumeCallback.() -> Unit) {
+        consumeCallback = callback
+
+        getNewIntentForBroadcast().apply {
+            action = ACTION_CONSUME
+            putExtra(KEY_TOKEN, purchaseToken)
+        }.run {
+            sendBroadcast(this)
+        }
+    }
+
+    override fun queryPurchasedProducts(
+        purchaseType: PurchaseType,
+        callback: PurchaseQueryCallback.() -> Unit
+    ) {
+        queryCallback = callback
+        getNewIntentForBroadcast().apply {
+            action = ACTION_QUERY_PURCHASES
+            putExtra(KEY_ITEM_TYPE, purchaseType.type)
+        }.run {
+            sendBroadcast(this)
+        }
+    }
+
+    override fun purchase(
+        activity: Activity,
+        purchaseRequest: PurchaseRequest,
+        purchaseType: PurchaseType,
+        callback: PurchaseIntentCallback.() -> Unit
+    ) {
+        purchaseActivityWeakReference = WeakReference(
+            PurchaseWeakHolder(activity, purchaseRequest.requestCode, callback)
+        )
+
+        sendPurchaseBroadcast(purchaseRequest, purchaseType, callback)
+    }
+
+    override fun purchase(
+        fragment: Fragment,
+        purchaseRequest: PurchaseRequest,
+        purchaseType: PurchaseType,
+        callback: PurchaseIntentCallback.() -> Unit
+    ) {
+        purchaseFragmentWeakReference = WeakReference(
+            PurchaseWeakHolder(fragment, purchaseRequest.requestCode, callback)
+        )
+
+        sendPurchaseBroadcast(purchaseRequest, purchaseType, callback)
+    }
+
+    private fun sendPurchaseBroadcast(
+        purchaseRequest: PurchaseRequest,
+        purchaseType: PurchaseType,
+        callback: PurchaseIntentCallback.() -> Unit
+    ) {
+        PurchaseIntentCallback().apply(callback).purchaseFlowBegan.invoke()
+        getNewIntentForBroadcast().apply {
+            action = ACTION_PURCHASE
+            putExtra(KEY_SKU, purchaseRequest.productId)
+            putExtra(KEY_DEVELOPER_PAYLOAD, purchaseRequest.payload)
+            putExtra(KEY_ITEM_TYPE, purchaseType.type)
+
+        }.run {
+            sendBroadcast(this)
+        }
+    }
+
+
+    override fun stopConnection() {
+        disconnected = true
+
+        cleanReferences()
+
+        receiverCommunicator?.let {
+            BillingReceiver.removeObserver(it)
+        }
+        receiverCommunicator = null
+    }
+
+    private fun cleanReferences() {
+        consumeCallback = null
+        queryCallback = null
+        connectionCallbackReference = null
+        contextReference = null
+
+        purchaseFragmentWeakReference = null
+        purchaseActivityWeakReference = null
```

Comment:
You should call `clear` on these WeakReferences.

Reply:
Done